frame/name-service/src/lib.rs

```diff
@@ -0,0 +1,351 @@
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A simple name service that can be used to give accounts friendly names.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use sp_std::prelude::*;
+use sp_runtime::traits::{Saturating};
+use frame_support::{decl_module, decl_error, decl_event, decl_storage, ensure, RuntimeDebug};
+use frame_support::dispatch::DispatchResult;
+use frame_support::traits::{
+	Currency, ReservableCurrency, Get, EnsureOrigin, OnUnbalanced,
+	WithdrawReason, ExistenceRequirement::KeepAlive, Imbalance,
+};
+use frame_system::ensure_signed;
+use codec::{Encode, Decode};
+
+mod mock;
+mod tests;
+mod benchmarking;
+
+type BalanceOf<T> = <<T as Trait>::Currency as Currency<<T as frame_system::Trait>::AccountId>>::Balance;
+type NegativeImbalanceOf<T> = <<T as Trait>::Currency as Currency<<T as frame_system::Trait>::AccountId>>::NegativeImbalance;
+
+pub trait WeightInfo {}
+
+/// The module's config trait.
+pub trait Trait: frame_system::Trait {
+
+	/// The currency trait.
+	type Currency: ReservableCurrency<Self::AccountId>;
+
+	/// The overarching event type.
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+
+	/// Origin that can have high level control over the name-service pallet.
+	type ManagerOrigin: EnsureOrigin<Self::Origin>;
+
+	/// Origin that can set permanent ownership of a name to an account.
+	type PermanenceOrigin: EnsureOrigin<Self::Origin>;
+
+	/// Time available between subsequent bids for a name.
+	type BiddingPeriod: Get<Self::BlockNumber>;
+
+	/// Time available after bidding has completed for the winner to claim their name.
+	type ClaimPeriod: Get<Self::BlockNumber>;
+
+	/// One ownership period, which can be multiplied through exponential deposit.
+	type OwnershipPeriod: Get<Self::BlockNumber>;
+
+	/// Handler for the unbalanced decrease when funds are burned.
+	type BurnDestination: OnUnbalanced<NegativeImbalanceOf<Self>>;
+
+	/// Minimum length for a publicly available name.
+	type MinLength: Get<usize>;
+
+	/// Maximum length for a publicly available name.
+	type MaxLength: Get<usize>;
+
+	/// Minimum Bid for a name.
+	type MinBid: Get<BalanceOf<Self>>;
+
+	/// Weight information for extrinsics in this pallet.
+	type WeightInfo: WeightInfo;
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug)]
+pub enum Name<AccountId, BlockNumber, Balance> {
+	Available,
+	Bidding {
+		who: AccountId,
+		expiration: BlockNumber,
```

Comment:
so this is when bidding expires, not when name expires? having both called `expiration` is confusing